File: app/src/main/java/com/jcons/pdfconverter/pdfconverter/PDFWriter.java
package com.jcons.pdfconverter.pdfconverter;

import android.graphics.Bitmap;

public class PDFWriter{

    private PDFDocument mDocument;
    private IndirectObject mCatalog;
    private Pages mPages;
    private Page mCurrentPage;

    public PDFWriter() {
        newDocument(PaperSize.A4_WIDTH, PaperSize.A4_HEIGHT);
    }

    private void newDocument(int pageWidth, int pageHeight) {
        mDocument = new PDFDocument();
        mCatalog = mDocument.newIndirectObject();
        mDocument.includeIndirectObject(mCatalog);
        mPages = new Pages(mDocument, pageWidth, pageHeight);
        mDocument.includeIndirectObject(mPages.getIndirectObject());
        renderCatalog();
        newPage();
    }

    private void renderCatalog() {
        mCatalog.setDictionaryContent("  /Type /Catalog\n  /Pages " + mPages.getIndirectObject().getIndirectReference() + "\n");
    }

    public void newPage() {
        mCurrentPage = mPages.newPage();
        mDocument.includeIndirectObject(mCurrentPage.getIndirectObject());
        mPages.render();
    }

    public void addImage(int fromLeft, int fromBottom, Bitmap bitmap) {
        addImage(fromLeft, fromBottom, bitmap, Transformation.DEGREES_0_ROTATION);
    }

    public void addImage(int fromLeft, int fromBottom, Bitmap bitmap, String transformation) {
        final XObjectImage xImage = new XObjectImage(mDocument, bitmap);
        mCurrentPage.addImage(fromLeft, fromBottom, xImage.getWidth(), xImage.getHeight(), xImage, transformation);
    }

    public String asString() {
        mPages.render();
        return mDocument.toPDFString();
    }

}

File: app/src/main/java/com/jcons/pdfconverter/pdfconverter/PDFWriter.kt
package com.jcons.pdfconverter.pdfconverter

import android.graphics.Bitmap


class PDFWriter {
    private var mDocument: PDFDocument = PDFDocument()
    private var mCatalog: IndirectObject? = null
    private var mPages: Pages? = null
    private var mCurrentPage: Page? = null

    private fun renderCatalog() {
        mCatalog!!.setDictionaryContent(
            """  /Type /Catalog
            /Pages ${mPages?.indirectObject?.indirectReference}
            """
        )
    }

    fun newPage() {
        mCurrentPage = mPages!!.newPage()
        mDocument!!.includeIndirectObject(mCurrentPage!!.indirectObject)
        mPages!!.render()
    }

    @JvmOverloads
    fun addImage(
        fromLeft: Int,
        fromBottom: Int,
        bitmap: Bitmap,
        transformation: String = DEGREES_0_ROTATION
    ) {
        val xImage = XObjectImage(mDocument, bitmap)
        mCurrentPage!!.addImage(
            fromLeft,
            fromBottom,
            xImage.width,
            xImage.height,
            xImage,
            transformation
        )
    }

    fun asString(): String {
        mPages!!.render()
        return mDocument!!.toPDFString()
    }

    companion object {
        const val DEGREES_0_ROTATION = "1 0 0 1"
    }

    init {
        mCatalog = mDocument!!.newIndirectObject()
        mDocument!!.includeIndirectObject(mCatalog)
        mPages = Pages(mDocument!!)
        mDocument!!.includeIndirectObject(mPages?.indirectObject)
        renderCatalog()
        newPage()
    }
}
